extensions/binding/org.eclipse.smarthome.binding.bluetooth/src/main/java/org/eclipse/smarthome/binding/bluetooth/discovery/internal/BluetoothDiscoveryService.java

```diff
@@ -0,0 +1,178 @@
+import org.osgi.service.component.annotations.ReferenceCardinality;
+import org.osgi.service.component.annotations.ReferencePolicy;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link BluetoothDiscoveryService} handles searching for BLE devices.
+ *
+ * @author Chris Jackson - Initial Contribution
+ * @author Kai Kreuzer - Introduced BluetoothAdapters and BluetoothDiscoveryParticipants
+ *
+ */
+@Component(immediate = true, service = DiscoveryService.class, configurationPid = "discovery.bluetooth")
+public class BluetoothDiscoveryService extends AbstractDiscoveryService {
+
+    private final Logger logger = LoggerFactory.getLogger(BluetoothDiscoveryService.class);
+
+    private final static int SEARCH_TIME = 15;
+
+    private final Set<BluetoothAdapter> adapters = new CopyOnWriteArraySet<>();
+    private final Set<BluetoothDiscoveryParticipant> participants = new CopyOnWriteArraySet<>();
+    private final Map<UID, BluetoothDiscoveryListener> registeredListeners = new HashMap<>();
+
+    public BluetoothDiscoveryService() {
+        super(SEARCH_TIME);
+    }
+
+    @Override
+    protected void activate(Map<String, Object> configProperties) {
+        logger.debug("Activating Bluetooth discovery service");
+        super.activate(configProperties);
+        startScan();
+    }
+
+    @Override
+    @Modified
+    protected void modified(Map<String, Object> configProperties) {
+        super.modified(configProperties);
+    }
+
+    @Override
+    public void deactivate() {
+        logger.debug("Deactivating Bluetooth discovery service");
+    }
+
+    @Reference(cardinality = ReferenceCardinality.MULTIPLE, policy = ReferencePolicy.DYNAMIC)
+    protected void addBluetoothAdapter(BluetoothAdapter adapter) {
+        this.adapters.add(adapter);
+        BluetoothDiscoveryListener listener = new BluetoothDiscoveryListener() {
+
+            @Override
+            public void deviceDiscovered(BluetoothDevice device) {
+                BluetoothDiscoveryService.this.deviceDiscovered(adapter, device);
+
+            }
+        };
+        adapter.addDiscoveryListener(listener);
+        registeredListeners.put(adapter.getUID(), listener);
+    }
+
+    protected void removeBluetoothAdapter(BluetoothAdapter adapter) {
+        this.adapters.remove(adapter);
+        adapter.removeDiscoveryListener(registeredListeners.remove(adapter.getUID()));
+    }
+
+    @Reference(cardinality = ReferenceCardinality.MULTIPLE, policy = ReferencePolicy.DYNAMIC)
+    protected void addBluetoothDiscoveryParticipant(BluetoothDiscoveryParticipant participant) {
+        this.participants.add(participant);
+    }
+
+    protected void removeBluetoothDiscoveryParticipant(BluetoothDiscoveryParticipant participant) {
+        this.participants.remove(participant);
+    }
+
+    @Override
+    public Set<ThingTypeUID> getSupportedThingTypes() {
+        Set<ThingTypeUID> supportedThingTypes = new HashSet<>();
```

Comment:
I _think_  it would make sense to construct this set whenever a BDP is added/removed and keep it ina a field, because IIRC this method gets called quite frequently.